### ISSUE TITLE
peek-poke is unsound

### ISSUE THREAD

**Issue by nox** [servo/webrender#3766]
Method `peek_from_uninit` creates a `&mut T` where the `T` is uninitialised, this is UB AFAIK.

https://github.com/servo/webrender/blob/efec27697590b622a4cc67af31c3d695f920f80f/peek-poke/src/lib.rs#L33-L39

**Comment by nox** [comment]
@SimonSapin says "I think the lang team is not decided yet whether &mut T is UB, when T is valid for any bit pattern", though.

**Comment by nox** [comment]
Cc @RalfJung

**Comment by RalfJung** [comment]
Right, so this is UB right now to keep our options open.

However, for the specific case of uninhabited `T`, we likely do want to actually make this UB "for real" as that would help greatly with dead-code removal. Does/could `!` implement `Peek`? If yes, this could would be UB and risk miscompilation in future compiler versions. So `Peek` might have to be an `unsafe trait` for this to be sound.

See https://github.com/rust-lang/unsafe-code-guidelines/issues/77 for the UCG discussion on this. If you think it is impossible to do what you are doing here when `&mut T` always requires a fully valid `T`, that would be an interesting data point for us.

**Comment by jrmuizel** [comment]
Thoughts @Gankra?

**Comment by Gankra** [comment]
Yeah it's "Documentation UB" in that it's reserved just to be conservative but it's pretty obvious this pattern needs to be *doable* somehow, so it's only UB insofar as the lang team has failed to design a mechanism that they're happy with for expressing this pattern.

Which is to say, this code is fine today. I maintain the docs that say this is UB, and I reviewed the code, so this code was accepted with full awareness of the situation. Ralf is also aware of my stance on this problem from the discussion we've had on the aforementionned docs.

Some day, the lang team might either:

* Decide it's fine (no changes needed)
* Design a new thing to express this pattern (like &raw or offsetof) and then we can update the code to use that.

Note that I made peek_from only work with `*mut T` so this is literally the &raw/offsetof usecase (transiently creating an &mut simply because we lack another way to compute an address).

**Comment by RalfJung** [comment]
> Note that I made peek_from only work with *mut T so this is literally the &raw/offsetof usecase (transiently creating an &mut simply because we lack another way to compute an address).

Ah, that wasn't clear. (Sometimes type inference isn't helping... like when reading a snippet of 3 lines of code without having any context.^^)

But then, why doesn't this work?
```rust
     let mut val = MaybeUninit::uninit(); 
     let bytes = <T>::peek_from(bytes, val.as_mut_ptr()); 
     (val.assume_init(), bytes) 
```

**Comment by Gankra** [comment]
That's just the same code hidden behind a function call :)

**Comment by RalfJung** [comment]
Yes and no. Code in libstd is special and may make assumptions that code outside libstd may not.

But also, for both of them, there's actually no reason to wait for `&raw`: You could just do
```rust
 pub unsafe fn peek_from_uninit<T: Copy + Peek>(bytes: *const u8) -> (T, *const u8) { 
     let mut val = MaybeUninitShim { uninit: () }; 
     let bytes = <T>::peek_from(bytes, &mut val as *mut _ as *mut T); 
     (val.init, bytes) 
 } 
```

**Comment by Gankra** [comment]
The code predates MaybeUninit's stabilization, so yeah we could reasonably update to use it so we can "forget" about the issue and have it solved for us later.

Just not a big deal in my mind.
